File: jas_dioxus/src/tools/anchor_point_tool.rs
//! Anchor Point (Convert Anchor Point) tool.
//!
//! Three interactions:
//! - Drag on a corner point: pull out symmetric control handles (→ smooth).
//! - Click on a smooth point: collapse handles to anchor (→ corner).
//! - Drag on a control handle: move that handle independently (→ cusp).

use web_sys::CanvasRenderingContext2d;

use crate::document::document::ElementSelection;
use crate::document::model::Model;
use crate::geometry::element::{
    control_points, convert_corner_to_smooth, convert_smooth_to_corner,
    is_smooth_point, move_path_handle_independent, path_handle_positions, Element, PathElem,
};

use super::tool::{CanvasTool, HIT_RADIUS};

#[derive(Debug, Clone)]
enum State {
    Idle,
    /// Dragging from a corner anchor to pull out handles.
    DraggingCorner {
        path: Vec<usize>,
        pe: PathElem,
        anchor_idx: usize,
        start_x: f64,
        start_y: f64,
    },
    /// Dragging a control handle independently (cusp).
    DraggingHandle {
        path: Vec<usize>,
        pe: PathElem,
        anchor_idx: usize,
        handle_type: String,
        start_hx: f64,
        start_hy: f64,
    },
    /// Pressed on a smooth point — will convert to corner on release if no drag.
    PressedSmooth {
        path: Vec<usize>,
        pe: PathElem,
        anchor_idx: usize,
        start_x: f64,
        start_y: f64,
    },
}

pub struct AnchorPointTool {
    state: State,
}

impl AnchorPointTool {
    pub fn new() -> Self {
        Self { state: State::Idle }
    }

    /// Hit-test bezier handles on all path elements (not just selected ones).
    fn hit_test_handle(
        model: &Model, x: f64, y: f64,
    ) -> Option<(Vec<usize>, PathElem, usize, String, f64, f64)> {
        let doc = model.document();
        for (li, layer) in doc.layers.iter().enumerate() {
            if let Some(children) = layer.children() {
                for (ci, child) in children.iter().enumerate() {
                    if let Element::Path(pe) = &**child
                        && let Some(result) = check_handles(pe, &[li, ci], x, y) {
                            return Some(result);
                        }
                    if let Element::Group(g) = &**child {
                        if child.common().locked { continue; }
                        for (gi, gc) in g.children.iter().enumerate() {
                            if let Element::Path(pe) = &**gc
                                && let Some(result) = check_handles(pe, &[li, ci, gi], x, y) {
                                    return Some(result);
                                }
                        }
                    }
                }
            }
        }
        None
    }

    /// Hit-test anchor points on all path elements.
    fn hit_test_anchor(
        model: &Model, x: f64, y: f64,
    ) -> Option<(Vec<usize>, PathElem, usize)> {
        let doc = model.document();
        for (li, layer) in doc.layers.iter().enumerate() {
            if let Some(children) = layer.children() {
                for (ci, child) in children.iter().enumerate() {
                    if let Element::Path(pe) = &**child
                        && let Some(idx) = find_anchor_at(pe, x, y) {
                            return Some((vec![li, ci], pe.clone(), idx));
                        }
                    if let Element::Group(g) = &**child {
                        if child.common().locked { continue; }
                        for (gi, gc) in g.children.iter().enumerate() {
                            if let Element::Path(pe) = &**gc
                                && let Some(idx) = find_anchor_at(pe, x, y) {
                                    return Some((vec![li, ci, gi], pe.clone(), idx));
                                }
                        }
                    }
                }
            }
        }
        None
    }

    fn select_all_cps(model: &mut Model, path: &[usize], _elem: &Element) {
        let mut doc = model.document().clone();
        doc.selection.retain(|es| es.path != path);
        doc.selection.push(ElementSelection::all(path.to_vec()));
        model.set_document(doc);
    }
}

fn check_handles(
    pe: &PathElem, path: &[usize], x: f64, y: f64,
) -> Option<(Vec<usize>, PathElem, usize, String, f64, f64)> {
    let anchors = control_points(&Element::Path(pe.clone()));
    for (ai, _) in anchors.iter().enumerate() {
        let (h_in, h_out) = path_handle_positions(&pe.d, ai);
        if let Some((hx, hy)) = h_in
            && ((x - hx).powi(2) + (y - hy).powi(2)).sqrt() < HIT_RADIUS {
                return Some((path.to_vec(), pe.clone(), ai, "in".to_string(), hx, hy));
            }
        if let Some((hx, hy)) = h_out
            && ((x - hx).powi(2) + (y - hy).powi(2)).sqrt() < HIT_RADIUS {
                return Some((path.to_vec(), pe.clone(), ai, "out".to_string(), hx, hy));
            }
    }
    None
}

fn find_anchor_at(pe: &PathElem, px: f64, py: f64) -> Option<usize> {
    let anchors = control_points(&Element::Path(pe.clone()));
    for (i, &(ax, ay)) in anchors.iter().enumerate() {
        if ((px - ax).powi(2) + (py - ay).powi(2)).sqrt() < HIT_RADIUS {
            return Some(i);
        }
    }
    None
}

impl CanvasTool for AnchorPointTool {
    fn on_press(&mut self, model: &mut Model, x: f64, y: f64, _shift: bool, _alt: bool) {
        // 1. Check handle hit first (for cusp behavior)
        if let Some((path, pe, anchor_idx, handle_type, hx, hy)) =
            Self::hit_test_handle(model, x, y)
        {
            self.state = State::DraggingHandle {
                path,
                pe,
                anchor_idx,
                handle_type,
                start_hx: hx,
                start_hy: hy,
            };
            return;
        }

        // 2. Check anchor point hit
        if let Some((path, pe, anchor_idx)) = Self::hit_test_anchor(model, x, y) {
            if is_smooth_point(&pe.d, anchor_idx) {
                // Smooth point — will convert to corner on click, or ignore drag
                self.state = State::PressedSmooth {
                    path,
                    pe,
                    anchor_idx,
                    start_x: x,
                    start_y: y,
                };
            } else {
                // Corner point — start dragging to pull out handles
                self.state = State::DraggingCorner {
                    path,
                    pe,
                    anchor_idx,
                    start_x: x,
                    start_y: y,
                };
            }
        }
    }

    fn on_move(&mut self, model: &mut Model, x: f64, y: f64, _shift: bool, _alt: bool, dragging: bool) {
        if !dragging {
            return;
        }

        match &self.state {
            State::DraggingCorner { path, pe, anchor_idx, .. } => {
                // Live preview: convert corner to smooth with handles at drag position
                let new_pe = convert_corner_to_smooth(pe, *anchor_idx, x, y);
                let new_elem = Element::Path(new_pe);
                let doc = model.document().replace_element(path, new_elem.clone());
                model.set_document(doc);
            }
            State::DraggingHandle { path, pe, anchor_idx, handle_type, start_hx, start_hy } => {
                // Move handle independently (cusp)
                let dx = x - start_hx;
                let dy = y - start_hy;
                let new_pe = move_path_handle_independent(pe, *anchor_idx, handle_type, dx, dy);
                let new_elem = Element::Path(new_pe);
                let doc = model.document().replace_element(path, new_elem.clone());
                model.set_document(doc);
            }
            State::PressedSmooth { start_x, start_y, path, pe, anchor_idx } => {
                // If dragged far enough from a smooth point, start pulling handles
                // (convert the interaction to a corner-drag)
                let dist = ((x - start_x).powi(2) + (y - start_y).powi(2)).sqrt();
                if dist > 3.0 {
                    // Convert to corner first, then start pulling new handles
                    let corner_pe = convert_smooth_to_corner(pe, *anchor_idx);
                    let new_pe = convert_corner_to_smooth(&corner_pe, *anchor_idx, x, y);
                    let new_elem = Element::Path(new_pe);
                    let doc = model.document().replace_element(path, new_elem.clone());
                    model.set_document(doc);

                    self.state = State::DraggingCorner {
                        path: path.clone(),
                        pe: corner_pe,
                        anchor_idx: *anchor_idx,
                        start_x: *start_x,
                        start_y: *start_y,
                    };
                }
            }
            State::Idle => {}
        }
    }

    fn on_release(&mut self, model: &mut Model, _x: f64, _y: f64, _shift: bool, _alt: bool) {
        let state = std::mem::replace(&mut self.state, State::Idle);
        match state {
            State::PressedSmooth { path, pe, anchor_idx, .. } => {
                // Click on smooth point → convert to corner
                model.snapshot();
                let new_pe = convert_smooth_to_corner(&pe, anchor_idx);
                let new_elem = Element::Path(new_pe.clone());
                let doc = model.document().replace_element(&path, new_elem.clone());
                model.set_document(doc);
                Self::select_all_cps(model, &path, &new_elem);
            }
            State::DraggingCorner { path, pe, anchor_idx, start_x, start_y, .. } => {
                let dist = ((_x - start_x).powi(2) + (_y - start_y).powi(2)).sqrt();
                if dist > 1.0 {
                    // Commit the corner→smooth conversion
                    model.snapshot();
                    let new_pe = convert_corner_to_smooth(&pe, anchor_idx, _x, _y);
                    let new_elem = Element::Path(new_pe);
                    let doc = model.document().replace_element(&path, new_elem.clone());
                    model.set_document(doc);
                    Self::select_all_cps(model, &path, &new_elem);
                }
            }
            State::DraggingHandle { path, pe, anchor_idx, handle_type, start_hx, start_hy } => {
                let dx = _x - start_hx;
                let dy = _y - start_hy;
                if dx.abs() > 0.5 || dy.abs() > 0.5 {
                    model.snapshot();
                    let new_pe = move_path_handle_independent(&pe, anchor_idx, &handle_type, dx, dy);
                    let new_elem = Element::Path(new_pe);
                    let doc = model.document().replace_element(&path, new_elem.clone());
                    model.set_document(doc);
                    Self::select_all_cps(model, &path, &new_elem);
                }
            }
            State::Idle => {}
        }
    }

    fn draw_overlay(&self, _model: &Model, _ctx: &CanvasRenderingContext2d) {
        // No overlay needed
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::geometry::element::PathCommand;

    fn make_line_path() -> PathElem {
        PathElem {
            d: vec![
                PathCommand::MoveTo { x: 0.0, y: 0.0 },
                PathCommand::LineTo { x: 50.0, y: 0.0 },
                PathCommand::LineTo { x: 100.0, y: 0.0 },
            ],
            fill: None,
            stroke: None,
            common: Default::default(),
        }
    }

    fn make_smooth_path() -> PathElem {
        PathElem {
            d: vec![
                PathCommand::MoveTo { x: 0.0, y: 0.0 },
                PathCommand::CurveTo { x1: 10.0, y1: 20.0, x2: 40.0, y2: 20.0, x: 50.0, y: 0.0 },
                PathCommand::CurveTo { x1: 60.0, y1: -20.0, x2: 90.0, y2: -20.0, x: 100.0, y: 0.0 },
            ],
            fill: None,
            stroke: None,
            common: Default::default(),
        }
    }

    #[test]
    fn corner_point_is_not_smooth() {
        let pe = make_line_path();
        assert!(!is_smooth_point(&pe.d, 0)); // MoveTo
        assert!(!is_smooth_point(&pe.d, 1)); // LineTo - no handles
        assert!(!is_smooth_point(&pe.d, 2)); // LineTo - no handles
    }

    #[test]
    fn smooth_point_is_smooth() {
        let pe = make_smooth_path();
        // Anchor at index 1 (50,0) has incoming handle (40,20) and outgoing (60,-20)
        assert!(is_smooth_point(&pe.d, 1));
    }

    #[test]
    fn convert_corner_to_smooth_creates_handles() {
        let pe = make_line_path();
        // Convert anchor 1 (at 50,0) to smooth by dragging to (50, 30)
        let result = convert_corner_to_smooth(&pe, 1, 50.0, 30.0);
        // Should have CurveTo commands now
        assert!(matches!(result.d[1], PathCommand::CurveTo { .. }));
        // Outgoing handle on next segment should be at (50,30)
        if let PathCommand::CurveTo { x1, y1, .. } = result.d[2] {
            assert!((x1 - 50.0).abs() < 0.01);
            assert!((y1 - 30.0).abs() < 0.01);
        } else {
            panic!("expected CurveTo for segment after converted point");
        }
        // Incoming handle should be reflected: (50, -30)
        if let PathCommand::CurveTo { x2, y2, .. } = result.d[1] {
            assert!((x2 - 50.0).abs() < 0.01);
            assert!((y2 - (-30.0)).abs() < 0.01);
        } else {
            panic!("expected CurveTo");
        }
    }

    #[test]
    fn convert_smooth_to_corner_collapses_handles() {
        let pe = make_smooth_path();
        let result = convert_smooth_to_corner(&pe, 1);
        // After conversion, anchor 1 should have no visible handles
        assert!(!is_smooth_point(&result.d, 1));
        // The x2,y2 of cmd[1] should equal the anchor (50,0)
        if let PathCommand::CurveTo { x2, y2, x, y, .. } = result.d[1] {
            assert!((x2 - x).abs() < 0.01);
            assert!((y2 - y).abs() < 0.01);
        }
        // The x1,y1 of cmd[2] should equal the anchor (50,0)
        if let PathCommand::CurveTo { x1, y1, .. } = result.d[2] {
            assert!((x1 - 50.0).abs() < 0.01);
            assert!((y1 - 0.0).abs() < 0.01);
        }
    }

    #[test]
    fn independent_handle_move_does_not_reflect() {
        let pe = make_smooth_path();
        // Move the outgoing handle of anchor 1 by (10, 5)
        let result = move_path_handle_independent(&pe, 1, "out", 10.0, 5.0);
        // Outgoing handle (x1 of cmd[2]) should be moved
        if let PathCommand::CurveTo { x1, y1, .. } = result.d[2] {
            assert!((x1 - 70.0).abs() < 0.01); // 60 + 10
            assert!((y1 - (-15.0)).abs() < 0.01); // -20 + 5
        }
        // Incoming handle (x2 of cmd[1]) should NOT be changed
        if let PathCommand::CurveTo { x2, y2, .. } = result.d[1] {
            assert!((x2 - 40.0).abs() < 0.01); // unchanged
            assert!((y2 - 20.0).abs() < 0.01); // unchanged
        }
    }

    #[test]
    fn find_anchor_at_hits_near_point() {
        let pe = make_line_path();
        // Point near anchor 1 (50, 0)
        assert_eq!(find_anchor_at(&pe, 51.0, 1.0), Some(1));
        // Point far from any anchor
        assert_eq!(find_anchor_at(&pe, 200.0, 200.0), None);
    }

    #[test]
    fn convert_first_anchor_corner_to_smooth() {
        let pe = make_line_path();
        // Convert anchor 0 (MoveTo at 0,0) — only outgoing handle
        let result = convert_corner_to_smooth(&pe, 0, 10.0, 20.0);
        // The next command should get the outgoing handle
        if let PathCommand::CurveTo { x1, y1, .. } = result.d[1] {
            assert!((x1 - 10.0).abs() < 0.01);
            assert!((y1 - 20.0).abs() < 0.01);
        } else {
            panic!("expected CurveTo after converting first anchor");
        }
    }

    #[test]
    fn convert_last_anchor_corner_to_smooth() {
        let pe = make_line_path();
        // Convert anchor 2 (last, LineTo at 100,0) — only incoming handle
        let result = convert_corner_to_smooth(&pe, 2, 100.0, 30.0);
        // The incoming handle (x2,y2 on this cmd) should be the reflected position
        if let PathCommand::CurveTo { x2, y2, x, y, .. } = result.d[2] {
            // Reflected of (100,30) through (100,0) = (100,-30)
            assert!((x2 - 100.0).abs() < 0.01);
            assert!((y2 - (-30.0)).abs() < 0.01);
            assert!((x - 100.0).abs() < 0.01);
            assert!((y - 0.0).abs() < 0.01);
        } else {
            panic!("expected CurveTo");
        }
    }
}

File: jas_ocaml/lib/tools/anchor_point_tool.ml
(** Anchor Point (Convert) tool.

    Three interactions, mirroring jas_dioxus/src/tools/anchor_point_tool.rs:

    - Drag on a corner anchor: pull out symmetric handles → smooth.
    - Click on a smooth anchor: collapse handles to anchor → corner.
    - Drag on a control handle: move it independently → cusp.

    Hit-test priority: handles before anchors. *)

let hit_radius = Canvas_tool.hit_radius

(** Anchor positions of a path, indexed by anchor index (closePath skipped). *)
let anchor_points d =
  List.fold_left (fun acc cmd ->
    match cmd with
    | Element.MoveTo (x, y) | Element.LineTo (x, y) -> (x, y) :: acc
    | Element.CurveTo (_, _, _, _, x, y) -> (x, y) :: acc
    | Element.SmoothCurveTo (_, _, x, y) -> (x, y) :: acc
    | Element.QuadTo (_, _, x, y) -> (x, y) :: acc
    | Element.SmoothQuadTo (x, y) -> (x, y) :: acc
    | Element.ArcTo (_, _, _, _, _, x, y) -> (x, y) :: acc
    | Element.ClosePath -> acc
  ) [] d |> List.rev

(** Find the anchor index of a point near (px, py). *)
let find_anchor_at_idx d px py =
  let pts = anchor_points d in
  let result = ref None in
  List.iteri (fun i (ax, ay) ->
    if !result = None then begin
      let dist = sqrt ((px -. ax) *. (px -. ax) +. (py -. ay) *. (py -. ay)) in
      if dist < hit_radius then result := Some i
    end
  ) pts;
  !result

(** Find a handle ((anchor_idx, handle_type, hx, hy)) near (px, py). *)
let find_handle_at d px py =
  let n_anchors = List.length (anchor_points d) in
  let result = ref None in
  let i = ref 0 in
  while !result = None && !i < n_anchors do
    let (h_in, h_out) = Element.path_handle_positions d !i in
    (match h_in with
     | Some (hx, hy) when sqrt ((px -. hx) *. (px -. hx) +. (py -. hy) *. (py -. hy)) < hit_radius ->
       result := Some (!i, "in", hx, hy)
     | _ -> ());
    if !result = None then
      (match h_out with
       | Some (hx, hy) when sqrt ((px -. hx) *. (px -. hx) +. (py -. hy) *. (py -. hy)) < hit_radius ->
         result := Some (!i, "out", hx, hy)
       | _ -> ());
    incr i
  done;
  !result

(** Walk every Path element in the document one level into unlocked
    groups, returning the first one for which `f` produces Some. *)
let each_path_element doc f =
  let result = ref None in
  Array.iteri (fun li layer ->
    if !result = None then begin
      let children = match layer with
        | Element.Layer { children; _ } -> children
        | _ -> [||]
      in
      Array.iteri (fun ci child ->
        if !result = None then begin
          (match child with
           | Element.Path { d; _ } ->
             (match f d with
              | Some r -> result := Some ([li; ci], child, r)
              | None -> ())
           | Element.Group { children = gc; locked; _ } when not locked ->
             Array.iteri (fun gi gchild ->
               if !result = None then
                 match gchild with
                 | Element.Path { d; _ } ->
                   (match f d with
                    | Some r -> result := Some ([li; ci; gi], gchild, r)
                    | None -> ())
                 | _ -> ()
             ) gc
           | _ -> ())
        end
      ) children
    end
  ) doc.Document.layers;
  !result

(** Replace a path element's command list, preserving all other fields. *)
let replace_path_cmds elem new_cmds =
  match elem with
  | Element.Path { fill; stroke; width_points; opacity; transform; locked; visibility; blend_mode; _ } ->
    Element.Path { d = new_cmds; fill; stroke; width_points; opacity; transform; locked; visibility; blend_mode; mask = None }
  | _ -> elem

(** Apply new commands to the path at `path` and push back into the model. *)
let push_path_update (ctx : Canvas_tool.tool_context) path elem new_cmds =
  let new_elem = replace_path_cmds elem new_cmds in
  let new_doc = Document.replace_element ctx.model#document path new_elem in
  ctx.model#set_document new_doc;
  ctx.request_update ()

(** Set the selection to "all CPs of `path`". *)
let select_all_cps (ctx : Canvas_tool.tool_context) path =
  let doc = ctx.model#document in
  let new_sel = Document.PathMap.add path
    (Document.element_selection_all path)
    doc.Document.selection in
  ctx.model#set_document { doc with Document.selection = new_sel }

(* ------------------------------------------------------------------ *)
(* Tool state                                                          *)
(* ------------------------------------------------------------------ *)

type state =
  | Idle
  | Dragging_corner of {
      path : int list;
      pe : Element.element;
      anchor_idx : int;
      start_x : float;
      start_y : float;
    }
  | Dragging_handle of {
      path : int list;
      pe : Element.element;
      anchor_idx : int;
      handle_type : string;
      start_hx : float;
      start_hy : float;
    }
  | Pressed_smooth of {
      path : int list;
      pe : Element.element;
      anchor_idx : int;
      start_x : float;
      start_y : float;
    }

class anchor_point_tool = object (_self)
  inherit Canvas_tool.default_methods
  val mutable state : state = Idle

  method on_press (ctx : Canvas_tool.tool_context) x y ~(shift : bool) ~(alt : bool) =
    ignore shift; ignore alt;
    let doc = ctx.model#document in
    (* 1. Handle hit takes priority (cusp behaviour). *)
    let handle_hit = each_path_element doc (fun d -> find_handle_at d x y) in
    match handle_hit with
    | Some (path, elem, (ai, ht, hx, hy)) ->
      state <- Dragging_handle {
        path; pe = elem; anchor_idx = ai;
        handle_type = ht; start_hx = hx; start_hy = hy;
      }
    | None ->
      (* 2. Anchor hit. *)
      let anchor_hit = each_path_element doc (fun d -> find_anchor_at_idx d x y) in
      match anchor_hit with
      | None -> ()
      | Some (path, elem, ai) ->
        let d = match elem with Element.Path { d; _ } -> d | _ -> [] in
        if Element.is_smooth_point d ai then
          state <- Pressed_smooth {
            path; pe = elem; anchor_idx = ai; start_x = x; start_y = y;
          }
        else
          state <- Dragging_corner {
            path; pe = elem; anchor_idx = ai; start_x = x; start_y = y;
          }

  method on_move (ctx : Canvas_tool.tool_context) x y ~(shift : bool) ~(dragging : bool) =
    ignore shift;
    if not dragging then ()
    else match state with
    | Dragging_corner { path; pe; anchor_idx; _ } ->
      let d = match pe with Element.Path { d; _ } -> d | _ -> [] in
      let new_cmds = Element.convert_corner_to_smooth d anchor_idx x y in
      push_path_update ctx path pe new_cmds
    | Dragging_handle { path; pe; anchor_idx; handle_type; start_hx; start_hy } ->
      let d = match pe with Element.Path { d; _ } -> d | _ -> [] in
      let dx = x -. start_hx and dy = y -. start_hy in
      let new_cmds = Element.move_path_handle_independent d anchor_idx handle_type dx dy in
      push_path_update ctx path pe new_cmds
    | Pressed_smooth { path; pe; anchor_idx; start_x; start_y } ->
      let dist = sqrt ((x -. start_x) *. (x -. start_x) +. (y -. start_y) *. (y -. start_y)) in
      if dist > 3.0 then begin
        let d = match pe with Element.Path { d; _ } -> d | _ -> [] in
        let corner_cmds = Element.convert_smooth_to_corner d anchor_idx in
        let new_cmds = Element.convert_corner_to_smooth corner_cmds anchor_idx x y in
        push_path_update ctx path pe new_cmds;
        let corner_pe = replace_path_cmds pe corner_cmds in
        state <- Dragging_corner {
          path; pe = corner_pe; anchor_idx; start_x; start_y;
        }
      end
    | Idle -> ()

  method on_release (ctx : Canvas_tool.tool_context) x y ~(shift : bool) ~(alt : bool) =
    ignore shift; ignore alt;
    let s = state in
    state <- Idle;
    match s with
    | Pressed_smooth { path; pe; anchor_idx; _ } ->
      ctx.model#snapshot;
      let d = match pe with Element.Path { d; _ } -> d | _ -> [] in
      let new_cmds = Element.convert_smooth_to_corner d anchor_idx in
      push_path_update ctx path pe new_cmds;
      select_all_cps ctx path
    | Dragging_corner { path; pe; anchor_idx; start_x; start_y } ->
      let dist = sqrt ((x -. start_x) *. (x -. start_x) +. (y -. start_y) *. (y -. start_y)) in
      if dist > 1.0 then begin
        ctx.model#snapshot;
        let d = match pe with Element.Path { d; _ } -> d | _ -> [] in
        let new_cmds = Element.convert_corner_to_smooth d anchor_idx x y in
        push_path_update ctx path pe new_cmds;
        select_all_cps ctx path
      end
    | Dragging_handle { path; pe; anchor_idx; handle_type; start_hx; start_hy } ->
      let dx = x -. start_hx and dy = y -. start_hy in
      if abs_float dx > 0.5 || abs_float dy > 0.5 then begin
        ctx.model#snapshot;
        let d = match pe with Element.Path { d; _ } -> d | _ -> [] in
        let new_cmds = Element.move_path_handle_independent d anchor_idx handle_type dx dy in
        push_path_update ctx path pe new_cmds;
        select_all_cps ctx path
      end
    | Idle -> ()

  method on_double_click (_ctx : Canvas_tool.tool_context) (_x : float) (_y : float) = ()

  method on_key (_ctx : Canvas_tool.tool_context) (_key : int) = false

  method on_key_release (_ctx : Canvas_tool.tool_context) (_key : int) = false

  method draw_overlay (_ctx : Canvas_tool.tool_context) (_cr : Cairo.context) = ()

  method activate (_ctx : Canvas_tool.tool_context) = ()

  method deactivate (_ctx : Canvas_tool.tool_context) =
    state <- Idle
end
